osd,librados: add manifest, redirect

myoungwon opened the pull request:
As discussed with @liewegas (http://marc.info/?l=ceph-devel&m=149252183623701&w=2)
We want to implement extensible tier (which can support redirection, dedup, external system).

As a first step, this PR proposes a simple redirect (http://pad.ceph.com/p/deduplication_how_dedup_manifists, https://github.com/myoungwon/ceph/commit/47f3567a0d5894e9272bf3bcabc95eb71f736a8, https://github.com/myoungwon/ceph/commit/7b322975bec00c16f2c32c085e9961f47aee2098)

Signed-off-by: Myoungwon Oh omwmw@sk.com

liewegas commented on src/osd/osd_types.h:
```diff
@@ -3959,6 +3996,34 @@ static inline ostream& operator<<(ostream& out, const notify_info_t& n) {
+  void decode(bufferlist::iterator &bl);
+  void dump(Formatter *f) const;
+};
+WRITE_CLASS_ENCODER(object_manifest_t)
```
For this patch, please include just the object_manifest_t and object_info_t changes and leave off the extensible_mode, maybe_handle_*, mon, and other changes.

Also, for object_manifest_t, please add a static const char *get_type_name(int t) and const char *get_type_name() accessor that returns "none", "chunked", etc., and also add an friend ostream& operator<<(...).  And in the object_info_t operator<< add "if (has_manifest()) out << " " << manifest;"

liewegas commented on src/osd/osd_types.cc:
```diff
@@ -4450,6 +4450,7 @@ void object_info_t::encode(bufferlist& bl, uint64_t features) const
   ::encode(expected_object_size, bl);
   ::encode(expected_write_size, bl);
   ::encode(alloc_hint_flags, bl);
+  ::encode(manifest, bl);
```
this hunk shoud be in teh manifest_t patch

liewegas commented on src/osd/osd_types.cc:
```diff
@@ -4538,6 +4539,9 @@ void object_info_t::decode(bufferlist::iterator& bl)
   }
+  if (struct_v >= 17) {
+    ::decode(manifest, bl);
+  }
```
this too

liewegas commented on src/osd/PrimaryLogPG.h:
```diff
@@ -1352,7 +1347,7 @@ class PrimaryLogPG : public PG, public PGBackend::Listener {
   map<ceph_tid_t, ProxyWriteOpRef> proxywrite_ops;
 
-  void do_proxy_write(OpRequestRef op, const hobject_t& missing_oid);
+  void do_proxy_write(OpRequestRef op, const hobject_t& missing_oid, ObjectContextRef obc = NULL);
```
this should get squashed iwth teh other parts of the first patch that you are making use of

liewegas commented on src/test/librados/tier.cc:
```diff
@@ -2745,6 +2745,59 @@ TEST_F(LibRadosTwoPoolsPP, CachePin) {
+  // wait for maps to settle before next test
+  cluster.wait_for_latest_osdmap();
+}
+
```
perfect!

liewegas commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -4287,22 +4287,6 @@ static string list_entries(const T& m) {
   return s;
 }
 
-void PrimaryLogPG::maybe_object_in_extensible_tier(ObjectState& obs)
```
this whole patch should get squashed in too

myoungwon commented:
@liewegas Fixed. Can you review the code to proxy reads and writes to the redirect?

liewegas commented on src/osd/osd_types.cc:
```diff
@@ -4511,6 +4565,7 @@ void object_info_t::dump(Formatter *f) const
   f->dump_unsigned("expected_object_size", expected_object_size);
   f->dump_unsigned("expected_write_size", expected_write_size);
   f->dump_unsigned("alloc_hint_flags", alloc_hint_flags);
+  manifest.dump(f);
```
f->dump_object("manifest", maniftest);

liewegas commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -2096,6 +2096,13 @@ void PrimaryLogPG::do_op(OpRequestRef& op)
   }
 
+  if (obc.get() && obc->obs.exists && obc->obs.oi.has_manifest()) {
+    if (maybe_handle_extensible_tier(op,
```
maybe_handle_manifest ?

liewegas commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -2239,6 +2246,38 @@ void PrimaryLogPG::do_op(OpRequestRef& op)
+    return cache_result_t::HANDLED_PROXY;
+  case object_manifest_t::TYPE_CHUNKED:
+  default:
+    assert(0 == "unrecognized extensible_mode");
```
s/extensible_mode/manifest type/

liewegas commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -5703,6 +5703,47 @@ int PrimaryLogPG::do_osd_ops(OpContext *ctx, vector<OSDOp>& ops)
       }
       break;
 
+    case CEPH_OSD_OP_SET_REDIRECT:
```
For now let's disallow this on anyother other than the base tier.  so, if pg_pool_t::is_tier() is true then EINVAL?  i would rather not worry about redirects interacting with cache tiering until we've throught it through carefully.

myoungwon commented:
@liewegas Fixed. and the functional test passed.
* bugs
I also fixed two bugs.
  - scrub error (oi.size and data size are different)
    --> add oi.size = 0, oi.new_object
  - omap clear (ec does not support omap)
    --> use omap.clear (if pool support omap)

* RadosMoel test
  - Regarding second test case (stress test), does osd return EAGAIN int the case of assert-ver ? I think osd will return EINVAL. 
  - I added SetRedirectOp class (in RadosModel.h). However, I am not sure that i understood well. Do i need to add all cases in a class (case 1, case 2)? Can you give me specific example? (cli commands)

myoungwon commented:
@liewegas I used following command in order to test redirection,
./ceph_test_rados --op copy_from 1 --op set_redirect 1 --max-ops 40 --objects 10 --max-in-flight 1 --size 4096 --max-seconds 60 --set_redirect

I have a few questions.
1. What if a redirected object (already set redirect) receives set_redirect command again? Do we need to send it to redirect_target? (I think an object's redirect_target need to be changed)
2. Can copy_from's source or target be redirected object? if so, there is a cyclic problem. (for example, oid 10 --> oid 9, oid 9 --> oid 10, send oid 10 a copy_from). Therefore, i think we need to return EINVAL if copy_from's target or source is redirected object)

liewegas commented:
1. Yeah, I think it should just change the manifest in that case.

2. Good point.  I think the best solution here is to do loop detection by checking when we get a copy-get op whether it is ultimately on behalf of ourselves.  I *think* the client id on the op is preserved when the read is proxied so we should be able to see whether it is us reading from ourselves and return EDEADLOCK or whatever...?

liewegas commented on src/test/osd/RadosModel.h:
```diff
@@ -1948,6 +1948,103 @@ class CopyFromOp : public TestOp {
+      context->oid_not_in_use.erase(oid_tgt);
+
+      // choose source snap
+      if (0 && !(rand() % 4) && !context->snaps.empty()) {
```
we can remove this branch since it's disabled.  and it's fine i think to ignore redirects to snaps.

liewegas commented on src/test/osd/RadosModel.h:
```diff
@@ -1948,6 +1948,103 @@ class CopyFromOp : public TestOp {
+    librados::IoCtx tgt_ctx;
+    context->rados.ioctx_create(oid_tgt.c_str(), tgt_ctx);
+    op.assert_version(src_value.version);
+    op.set_redirect(oid, tgt_ctx, 0);
```
This is incomplete.. there's no change to actually do this with some probability.  Missing a patch maybe?

In any case, I suggest start with the simpler case.  The thing with the redirect is if the target object content changes the src content object will change too, and RadosModel doesn't have anything to handle that sort of thing.  Instead, I think we should do something simpler: in the setup phase of the test, where we're writing the initial N objects, with some probability write the initial object to with a '-redirect-target-nnn suffix (or something unique) and make the real object a redirect to it.  Then no other changes.  That way we are doing all of the same operations we were doing before, with except that it would be proxied through a redirect.

That is much simpler (only need  to modify the setup phase; no need for a new op like this one).

Then, once that is working, we can add SetRedirect and UnsetRedirect (or whatever) operations that do the promote/demote type operations.  This part will be a bit trickier though since we want to make it race with other operations on the same object in order to exercise the race cases.  Let's make the simple one work first!

And put it in a separate commit from the unit test?

Thanks!

myoungwon commented:
@liewegas  Can you take a look this commit (https://github.com/ceph/ceph/pull/14894/commits/335adf1447bf057220fb29a27bfefbb7b8afb839)? I added a test case and the setup phase. 
Also, I passed the test (ceph_test_rados --op read 1 --op write 1  --max-ops 100 --objects 15 --max-in-flight 5 --size 4096 --set_redirect ).

And i have a few question.
1. set-redirect (version issue)
please look below scenario.
========
write oid 1 (version 1)
set-redirect (assert-ver 1, target oid 3 version 3)
oid 1's version is 3
copy_from (oid 4 (version 4) -> oid 1 (version 3))
oid 1's version is 4 on the client side, but actual version is not changed (still 3) because copy_from is redirected.
set-redirect (assert-ver 4, target oid 5 version 5)
now assert-ver will be failed
=========
How can we handle this? I think there are two solutions (1. unset-redirect first and update current version.  2. update version whenever proxy write is finished)

2. Do we need to add a new rados op (unset-redirect) for UnsetRedirect?

liewegas commented:
1.  pressure sure we don't want to write to the redirect on a proxied write; that would be expensive and also non-atomic.  I think unset before doing a new set-redirect makes sense.  If you do any normal read on the object you get the target data and user_version; we currently don't actually have a way to get the non-proxied version.  We probably need a IGNORE_REDIRECT op flag or something for that (analogous to the "l" in lstat, lchmod, etc.).

2. I can't tell if unset-redirect is any different from a delete w/ IGNORE_REDIRECT flag set.  I think it's not...?

Either way, I think IGNORE_REDIRECT is the next step.  The in your scenario above you can set that flag and stat the object to the redirect's version to use for the second set-redirect op.

myoungwon commented:
@liewegas Thanks for your comments. I created three commits based on your comments (https://github.com/ceph/ceph/pull/14894/commits/9fc12ea3c080c358827f6100469ca654972f69a9, https://github.com/ceph/ceph/pull/14894/commits/787d0dd0d299cd805f94a4671749afc4cc3aba64, https://github.com/ceph/ceph/pull/14894/commits/615aceb021e752c97f35f7fada16155a378886b9). Can you take a took?

And following tests are passed. (ceph_test_rados --op write 2 --op set_redirect 1 --op unset_redirect 1
--max-ops 100 --objects 20 --max-in-flight 1 --size 4096 --set_redirect, 
ceph_test_rados --op copy_from 1 --op set_redirect 1 --max-ops 100 --objects 20 --max-in-flight 1 --size 4096 --set_redirect )

myoungwon commented:
@liewegas Ping. Could you take a look at three new commits? (regarding IGNORE_REDIRECT, Unset-redirect)

liewegas commented on src/test/osd/RadosModel.h:
```diff
@@ -1948,6 +1966,197 @@ class CopyFromOp : public TestOp {
+	  cout << num << ":  got expected ENOENT (src dne)" << std::endl;
+	} else if (r == -ERANGE || r == -EOVERFLOW) {
+	  cout << num << ":  got expected ERANGE or EOVERFLOW ? v: " << src_value.version << std::endl;
+	  v_err = true;
```
can we verify here that is actually is expected?

liewegas commented on src/test/osd/RadosModel.h:
```diff
@@ -1948,6 +1966,197 @@ class CopyFromOp : public TestOp {
+	context->update_object_redirect_target(oid, oid_tgt);
+	if(!src_value.redirect_target.empty()) {
+	  context->oid_redirect_not_in_use.insert(src_value.redirect_target);
+	  context->oid_redirect_in_use.erase(src_value.redirect_target);
```
isn't this updated in _begin?

liewegas commented on src/test/osd/RadosModel.h:
```diff
@@ -1948,6 +1966,197 @@ class CopyFromOp : public TestOp {
+	  version = comp_stat_read->get_version64();
+	} else {
+	  cout << num << ": version is changed to v : " << comp_stat_read->get_version64();
+	  context->update_object_version(oid, comp_stat_read->get_version64());
```
I think we shouldn't need to update the model here; if it changed, it's because someone else did it, and they will update the model...?

myoungwon replied:
@liewegas These are for below test scenario. Do you think we don't need to care below case here? (new version is needed if version is mismatched (through stat()) as you mentioned..)
If so, i will remove this.

========
write oid 1 (version 1)
set-redirect (assert-ver 1, target oid 3 version 3)
oid 1's version is 3
copy_from (oid 4 (version 4) -> oid 1 (version 3))
oid 1's version is 4 on the client side, but actual version is not changed (still 3) because copy_from is redirected.
set-redirect (assert-ver 4, target oid 5 version 5)
now assert-ver will be failed --> (new version is needed)
=========

liewegas replied:
If I'm following, the time there is a mismatch is when we IGNORE_REDIRECT to do the set-redirect--that's the only time we care about the redirect version and not the target object version.  Can't we just precede any set-redirect with a stat w/ IGNORE_REDIRECT to get the redirect (vs logical object) version?

liewegas commented:
The radosmodel testing aside, this looks great!  I think we should add or patch a few cases in qa/suites/rados/thrash/workloads/*.yaml to enable redirects and then run this through qa!

myoungwon commented:
You are right. a stat w/ IGNORE_REDIRECT is right way. SetRedirect in RadosMoel.h has been reworked as follow.
- SetRedirect remove a object w/ IGNORE_REDIRECT and then create a new object if the object has a redirect target. This is because we need to synchronize object's version between source and target. Please look Case A as below. If oid 3 receive a copy_from(src is oid 1), this will be failed due to oid 1' version is different from oid 12 (oid 13 will send a copy_get (src is oid 1) to oid 1, but oid 12 receives copy_get from oid 1).
Case A
oid 1 (ver 1) ->  oid 12 (ver 2)   
oid 3 (ver 3) ->  oid 13 (ver 3)
- Stat before SetRedirect

And i think two test case can be added in qa/suites
1. ceph_test_rados --op read 100 --op write 100 --op copy_from 50 --op delete 50  --max-ops 4000 --objects 500  --set_redirect
2. ceph_test_rados --op copy_from 1 --op set_redirect 1 --max-ops 4000 --objects 500 - --set_redirect

Does that seems reasonable?

liewegas commented:
Why does the copy-get for osd 3 see osd 1's version instead of being proxied to osd 12?  It seems like it should only see the oid 1 version if it has the IGNORE_REDIRECT flag set?

liewegas commented:
Can you add a patch adding two yaml files to qa/suites/rados/thrash/workloads for those ceph_test_rados cases?

myoungwon commented:
I added two yaml files.
Regarding your question, Yes, please look case 1 -> 2 as below.

Case 1 
oid 1 (ver 1) -> oid 11 (ver 1)
oid 3 (ver 3) -> oid 13 (ver 3)

Case 2 (set-redirect (w/ IGNORE_DIRECT) oid 1 --> oid 12)
oid 1 (ver 1) -> oid 12 (ver 2)
oid 3 (ver 3) -> oid 13 (ver 3)

liewegas commented on src/test/osd/RadosModel.h:
```diff
@@ -1948,6 +1958,224 @@ class CopyFromOp : public TestOp {
+			    librados::OPERATION_ORDER_READS_WRITES,
+			    NULL);
+      comp->wait_for_safe();
+      context->update_object_version(oid_tgt, comp->get_version64());
```
could we have gotten this from our in-memory value of the version for oid?

myoungwon replied:
We use two object pools (A. objects for normal OP  B. objects for redirect target ).
Therefore, A indicate B.
In this case, copy_from will only update A's version (not B).

liewegas commented on src/test/osd/RadosModel.h:
```diff
@@ -1948,6 +1958,224 @@ class CopyFromOp : public TestOp {
+      context->update_object_full(oid, tgt_value);
+      context->update_object_version(oid, comp->get_version64());
+      comp->release();
+    }
```
This whole block could also do what UnsetRedirect does, right?  (Just remove the redirect?) 

(No need to change it; just making sure I understand the test.)

myoungwon replied:
Nope. UnsetRedirect just remove a object. But this block will not only remove a object but also create a new object (through copy_from).

liewegas commented:
Ok I think I'm following.  2 small questions on the test, and if you don't mind can you squash the RadosModel pieces into a single commit?  Then let's test this!

myoungwon commented:
Squash is done. Please review and test this.

myoungwon commented:
@liewegas 
http://pulpito.ceph.com/yuriw-2017-05-19_21:19:02-rados-wip-yuri-testing2_2017_5_20-distro-basic-smithi/

http://pulpito.ceph.com/yuriw-2017-05-20_04:22:19-rados-wip-yuri-testing2_2017_5_20---basic-smithi/

There are two kinds of errors. 
1. (a sequence error) the client send write requests to osd. (tid 1,2,3,4,5) But tid 5 is returned first.
2. (a version error) the client send copy_from to osd. But stat is returned first and then, copy_from's result is returned.

I thinks these errors seem to be caused by osd op order (although I can not check the osd logs because I'm not permitted yet). If an object (which has redirect manifest) receives any OP(write/read) requests, OP will be redirected without RW lock (like current proxy code).
Therefore, unordered OP will appear if failure (network down, osd killed) occur during the test.

Do we need to block RW ordered OP until prior OP is completed ? (or am i wrong?)

liewegas commented:
If you click the link in pulpito and remove teuthology.log from the URL you can see the full data set.  Look in remote/*/log/* to find the osd logs.

The proxy behavior is identical for cache pools so that doesn't sound like it would be the problem (or else we'd see these failures there, too).  We should be able to pipeline those proxied requests.

myoungwon commented:
@liewegas After checking the osd logs, I found the cause of these issues.
In ceph_test_rados, client send a read request (with RW ordered flag) in order to confirm object's version such as (write, read), (copy_from, read). However, current proxy code handles ordered read requests without adding CEPH_OSD_FLAG_RWORDERED. 

      if (op->may_write() || op->may_cache() || write_ordered) {
        if (can_proxy_write) {
          do_proxy_write(op, missing_oid);
          return cache_result_t::HANDLED_PROXY;
        }
      } else {
        do_proxy_read(op);
        return cache_result_t::HANDLED_PROXY;
      }

Above code is part of the proxy code. Read request with ordered flag can be handled by do_proxy_write() because write_ordered is true. And do_proxy_write() does not set CEPH_OSD_FLAG_RWORDERED other than do_proxy_read().
Therefore, below code (redirection) send a non-ordered read request without CEPH_OSD_FLAG_RWORDERED.

    if (op->may_write() || write_ordered) {
      do_proxy_write(op, obc->obs.oi.soid, obc);
    } else {
      do_proxy_read(op, obc);
    }
    return cache_result_t::HANDLED_PROXY;

This causes the client can receive the result of read before it receive write's result. (when target object enters recovery state).

So I think the current code should be changed as follows  (also proxy code).

    if ((op->may_write() || write_ordered) && !op->may_read()) {
      do_proxy_write(op, obc->obs.oi.soid, obc);
    } else {
      do_proxy_read(op, obc);
    }
    return cache_result_t::HANDLED_PROXY;

What do you think?

liewegas commented:
If the client asks for RWORDERED, we should honor that across the proxy.  I think the better fix is to make do_proxy_write add the RWORDERED flag when it needs to.  Something like
<pre>
diff --git a/src/osd/PrimaryLogPG.cc b/src/osd/PrimaryLogPG.cc
index 8c4f08c..85123d3 100644
--- a/src/osd/PrimaryLogPG.cc
+++ b/src/osd/PrimaryLogPG.cc
@@ -2859,6 +2859,9 @@ void PrimaryLogPG::do_proxy_write(OpRequestRef op, const hobject_t& missing_oid)
 
   const hobject_t& soid = m->get_hobj();
   unsigned flags = CEPH_OSD_FLAG_IGNORE_CACHE | CEPH_OSD_FLAG_IGNORE_OVERLAY;
+  if (!(op->may_write() || op->may_cache())) {
+    flags |= CEPH_OSD_FLAG_RWORDERED;
+  }
   dout(10) << __func__ << " Start proxy write for " << *m << dendl;
 
   ProxyWriteOpRef pwop(std::make_shared<ProxyWriteOp>(op, soid, m->ops, m->get_reqid()));
</pre>
?

yuriw commented:
tests tuns:

http://pulpito.ceph.com/yuriw-2017-05-19_21:19:02-rados-wip-yuri-testing2_2017_5_20-distro-basic-smithi/

http://pulpito.ceph.com/yuriw-2017-05-20_04:22:19-rados-wip-yuri-testing2_2017_5_20---basic-smithi/

myoungwon commented:
Thanks for your suggestion. Fixed.

liewegas commented:
myoungwon-2017-05-25_13:06:28-rados-wip-14894-myoungwon---basic-smithi

the failures are all unrelated. looks good!

liewegas commented on src/osd/PrimaryLogPG.cc:
```diff
@@ -5703,6 +5792,70 @@ int PrimaryLogPG::do_osd_ops(OpContext *ctx, vector<OSDOp>& ops)
+	if (!obs.exists) {
+	  result = -ENOENT;
+	  break;
+	}
```
Oops, one more thing we need here:  if (!get_osdmap()->test_flag(CEPH_OSDMAP_REQIURE_LUMINOUS)).  Otherwise this might get set on an upgrading cluster that doesn't fully support/understand the redirects.

myoungwon replied:
@liewegas Thanks for your comment. Fixed.

liewegas commented:
@jdurgin I think this is ready to merge.

I'm a little nervous about having this really basic support in luminous when it isn't all that useful yet (although maybe it is?).  We could guard set-redirect with a config option like 'osd_allow_redirect' for the time being until the capability is more complete.

That said, what we have here is complete AFAICS, with both function and stress tests.

liewegas commented:
Hmm I see a bunch of failures from the redirect tests in http://pulpito.ceph.com/sage-2017-05-26_17:02:41-rados-wip-sage-testing2---basic-smithi/

liewegas commented:
also,
<pre>
2017-05-26T21:22:52.203 INFO:tasks.workunit.client.0.smithi194.stdout:                 api_tier: [  FAILED  ] LibRadosTwoPoolsPP.SetRedirectRead
2017-05-26T21:22:52.203 INFO:tasks.workunit.client.0.smithi194.stdout:                 api_tier: [  FAILED  ] LibRadosTwoPoolsECPP.SetRedirectRead
</pre>
on /a/sage-2017-05-26_17:02:41-rados-wip-sage-testing2---basic-smithi/1232580

myoungwon commented:
@liewegas All failure seem to be due to lack of require_luminous flag (EOPNOTSUPP). Because "ceph osd set require_luminous_osd" is called later (in the middle of a test).

myoungwon commented:
@liewegas  In the success case (http://pulpito.ceph.com/sage-2017-05-26_17:02:41-rados-wip-sage-testing2---basic-smithi/1232585/), require_luminous is set at the beginning of the test. I think if we want to test redirect, we need to set require_luminous flag before run the test.

liewegas commented:
oh, right.  we'll need to get a little bit creative with the test suite then (i don't think we should blindly set the flag for the at-mkfs cases).   i can take it from here!

liewegas commented:
merged via #15325

myoungwon commented:
@liewegas I really appreciate your help. I will prepare the next step.

liewegas commented:
No problem!  This turned out great--thanks for taking the time to get the all of the tests in place.  I think it will pay off with the subsequent stages!

liewegas commented:
@myoungwon mind looking at http://tracker.ceph.com/issues/20114 ?

myoungwon commented:
@liewegas Sure. I will take a look it.

myoungwon commented:
@liewegas Sorry for my mistake. Failures are due to the following reason.
set-redirect test creates more objects than default due to redirect mapping (let's assume that we create 500 objects, oid 1 -> oid 501, oid 2 -> oid 502)
But the test should use objects within pre-defined range. For example, If the number of object is 500, only 1~500 oid are allowed for test operations.

In the failure cases, test operation tried to use wrong oids (which means higher than 500).
This is because the bug in the setup phase. If max_inflights > 1, set-redirect can be finished after setup is completed. (so wrong oid is included in oid_*_in_use).
We need to remove redirect objects in oid_*_in_use after setup is completed.
